source/extensions/path/uri_template_lib/uri_template_internal.cc

```diff
@@ -0,0 +1,386 @@
+    absl::FunctionRef<absl::StatusOr<ParsedResult<T>>(absl::string_view)> consume_func,
+    absl::string_view* patt) {
+
+  absl::StatusOr<ParsedResult<T>> status = consume_func(*patt);
+  if (!status.ok()) {
+    return status.status();
+  }
+  ParsedResult<T> result = *std::move(status);
+
+  *patt = result.unconsumed_pattern_;
+  return result.parsed_value_;
+}
+
+} // namespace
+
+std::string Variable::debugString() const { return toString(*this); }
+
+std::string ParsedUrlPattern::debugString() const {
+  return absl::StrCat("/", absl::StrJoin(parsed_segments_, "/", ToStringFormatter()),
+                      suffix_.value_or(""));
+}
+
+bool isValidLiteral(absl::string_view literal) {
+  static const std::string* kValidLiteralRegex =
+      new std::string(absl::StrCat("^[", kLiteral, "]+$"));
+  static const LazyRE2 literal_regex = {kValidLiteralRegex->data()};
+  return RE2::FullMatch(toStringPiece(literal), *literal_regex);
+}
+
+bool isValidRewriteLiteral(absl::string_view literal) {
+  static const std::string* kValidLiteralRegex =
+      new std::string(absl::StrCat("^[", kLiteral, "/]+$"));
+  static const LazyRE2 literal_regex = {kValidLiteralRegex->data()};
+  return RE2::FullMatch(toStringPiece(literal), *literal_regex);
+}
+
+bool isValidVariableName(absl::string_view indent) {
+  static const LazyRE2 ident_regex = {"^[a-zA-Z][a-zA-Z0-9_]*$"};
+  return RE2::FullMatch(toStringPiece(indent), *ident_regex);
+}
+
+absl::StatusOr<ParsedResult<Literal>> consumeLiteral(absl::string_view pattern) {
+  absl::string_view lit =
+      std::vector<absl::string_view>(absl::StrSplit(pattern, absl::MaxSplits('/', 1)))[0];
+  absl::string_view unconsumed_pattern = pattern.substr(lit.size());
+  if (!isValidLiteral(lit)) {
+    return absl::InvalidArgumentError("Invalid literal");
+  }
+  return ParsedResult<Literal>(lit, unconsumed_pattern);
+}
+
+absl::StatusOr<ParsedResult<Operator>> consumeOperator(absl::string_view pattern) {
+  if (absl::StartsWith(pattern, "**")) {
+    return ParsedResult<Operator>(Operator::TextGlob, pattern.substr(2));
+  }
+  if (absl::StartsWith(pattern, "*")) {
+    return ParsedResult<Operator>(Operator::PathGlob, pattern.substr(1));
+  }
+  return absl::InvalidArgumentError("Invalid Operator");
+}
+
+absl::StatusOr<ParsedResult<Variable>> consumeVariable(absl::string_view pattern) {
+  // Locate the variable pattern to parse.
+  if (pattern.size() < 2 || (pattern)[0] != '{') {
+    return absl::InvalidArgumentError("Invalid variable");
+  }
+  std::vector<absl::string_view> parts = absl::StrSplit(pattern.substr(1), absl::MaxSplits('}', 1));
+  if (parts.size() != 2) {
+    return absl::InvalidArgumentError("Unmatched variable bracket");
+  }
+  absl::string_view unconsumed_pattern = parts[1];
+
+  // Parse the actual variable pattern, starting with the variable name.
+  std::vector<absl::string_view> var_parts = absl::StrSplit(parts[0], absl::MaxSplits('=', 1));
+  if (!isValidVariableName(var_parts[0])) {
```

Comment:
nit: avoid abbreviation. Maybe "variable_parts" or just "parts"? (here and below for "var_")

Reply:
changed